File: solutions/erlang/sieve/1/src/sieve.erl
-module(sieve).

-export([primes/1]).

-define(LOWER_LIMIT, 2).
-define(COULD_BE_PRIME, true).
-define(NOT_PRIME, false).

primes(UpperLimit)
  when UpperLimit < ?LOWER_LIMIT ->
  [];

primes(UpperLimit) ->
  Range = lists:seq(?LOWER_LIMIT, UpperLimit),
  MarkedMap = create_marked_range(Range, ?COULD_BE_PRIME),
  PrimeMarkedMap = mark_primes(?LOWER_LIMIT, UpperLimit, MarkedMap),
  FilteredMap = could_be_prime(PrimeMarkedMap),
  maps:keys(FilteredMap).

mark_primes(nil, _UpperLimit, MarkedRange) ->
  MarkedRange;

mark_primes(CurrentValue, UpperLimit, OldMarkedMap) ->
  Disqualified = lists:seq(CurrentValue * 2, UpperLimit, CurrentValue),
  NewMarkedMap = create_marked_range(Disqualified, ?NOT_PRIME),
  MergedMarkedMap = maps:merge(OldMarkedMap, NewMarkedMap),
  NextValue = find_next_prime(CurrentValue, MergedMarkedMap),
  mark_primes(NextValue, UpperLimit, MergedMarkedMap).

create_marked_range(Range, Value) ->
  RangeMap = lists:map(fun(X) -> {X, Value} end, Range),
  maps:from_list(RangeMap).

could_be_prime(MarkedMap) ->
  maps:filter(fun could_be_prime/2, MarkedMap).

could_be_prime(_X, IsPrime) ->
  IsPrime.

find_next_prime(CurrentValue, MergedMarkedMap) ->
  PossiblePrimesMap = could_be_prime(MergedMarkedMap),
  PossiblePrimes = maps:keys(PossiblePrimesMap),
  LargerPossiblePrimes = lists:filter(fun(X) -> X > CurrentValue end, PossiblePrimes),
  find_next_prime(LargerPossiblePrimes).

find_next_prime([] = _LargerPossiblePrimes) ->
  nil;

find_next_prime(LargerPossiblePrimes) ->
  lists:min(LargerPossiblePrimes).

File: solutions/elixir/sieve/1/lib/sieve.ex
defmodule Sieve do
  @doc """
  Generates a list of primes up to a given limit.
  """

  @lower_limit 2
  @could_be_prime true
  @not_prime false

  @spec primes_to(non_neg_integer) :: [non_neg_integer]
  def primes_to(upper_limit)
      when upper_limit < @lower_limit,
      do: []

  def primes_to(upper_limit),
    do:
      @lower_limit..upper_limit
      |> create_marked_map(@could_be_prime)
      |> mark_primes(@lower_limit, upper_limit)
      |> filter_for_primes()
      |> Map.keys()
      |> Enum.sort()

  defp create_marked_map(range, is_prime),
    do:
      range
      |> Enum.map(&{&1, is_prime})
      |> Map.new()

  defp mark_primes(marked_map, nil, _upper_limit),
    do: marked_map

  defp mark_primes(old_marked_map, current_value, upper_limit) do
    new_marked_map =
      current_value
      |> create_disqualified_range(upper_limit)
      |> create_marked_map(@not_prime)

    merged_marked_map = Map.merge(old_marked_map, new_marked_map)

    next_value = find_next_possible_prime(merged_marked_map, current_value)
    mark_primes(merged_marked_map, next_value, upper_limit)
  end

  defp create_disqualified_range(current_value, upper_limit) do
    for x <- current_value..upper_limit,
        do: x * current_value
  end

  defp find_next_possible_prime(merged_marked_map, current_value),
    do:
      merged_marked_map
      |> filter_for_primes()
      |> Map.keys()
      |> Enum.filter(&(&1 > current_value))
      |> find_next_possible_prime()

  defp find_next_possible_prime([] = _larger_possible_primes),
    do: nil

  defp find_next_possible_prime(larger_possible_primes),
    do: Enum.min(larger_possible_primes)

  defp filter_for_primes(marked_map),
    do: :maps.filter(fn _, value -> value end, marked_map)
end
